src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java

```diff
@@ -438,4 +451,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
+        SimpleFeatureTypeBuilder tb = new SimpleFeatureTypeBuilder();
+        tb.setName("bar");
+        tb.add("name", String.class);
+        tb.add("geom", Point.class);
+
+        store.createSchema(tb.buildFeatureType());
+
+        CatalogBuilder cb = new CatalogBuilder(cat);
+        cb.setStore(ds);
+
+        SimpleFeatureStore fs = (SimpleFeatureStore) store.getFeatureSource("bar");
+
+        FeatureTypeInfo ft = cb.buildFeatureType(fs);
+        cat.add(ft);
+
+        String xml = "<wfs:Transaction service=\"WFS\" version=\"1.0.0\" "
+                + " xmlns:wfs=\"http://www.opengis.net/wfs\" "
+                + " xmlns:gml=\"http://www.opengis.net/gml\" "
+                + " xmlns:gs='" + SystemTestData.DEFAULT_URI + "'>"
+                + "<wfs:Insert idgen='UseExisting'>"
+                + " <gs:bar gml:id='00'>"
+                + "    <gs:name>acme</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='01'>"
+                + "    <gs:name>wiley</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='02'>"
+                + "    <gs:name>bugs</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='03'>"
+                + "    <gs:name>roadrunner</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='04'>"
+                + "    <gs:name>daffy</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='05'>"
+                + "    <gs:name>elmer</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='06'>"
+                + "    <gs:name>tweety</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='07'>"
+                + "    <gs:name>sylvester</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='08'>"
+                + "    <gs:name>marvin</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='09'>"
+                + "    <gs:name>yosemite</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='010'>"
+                + "    <gs:name>porky</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='011'>"
+                + "    <gs:name>speedy</gs:name>"
+                + " </gs:bar>"
+                + " <gs:bar gml:id='012'>"
+                + "    <gs:name>taz</gs:name>"
+                + " </gs:bar>"
+                + "</wfs:Insert>"
+                + "</wfs:Transaction>";
+
+        Document dom = postAsDOM("wfs", xml);
+        print(dom);
+        dom = getAsDOM("wfs?request=GetFeature&version=1.0.0&service=wfs&typeNames=gs:bar");
+        print(dom);
+
+        NodeList elementsByTagName = dom.getElementsByTagName("gs:bar");
+        for (int i=0; i<elementsByTagName.getLength(); i++) {
+                String id = elementsByTagName.item(i).getAttributes().item(0).getNodeValue();
+                assertEquals("bar."+i, id );
```

Comment:
This looks positive, does not look like your gml:id is being preserved - but your fix is working.

You can try changing your gml:id to "bar.00" ... "bar.012" (if you are saving into a datastore that can preserve fids).